In [1]:
import numpy as np
import tensorflow as tf

### Loading train, validation and test data

In [2]:
#creating a temporary variable npz,to store each of the three Employee_attrition datasets
npz = np.load('Employee_attrition_data_train.npz')

# extracting the inputs and making them floats
train_inputs = npz['inputs'].astype(np.float)

# casting targets to int for the sparse_categorical_crossentropy to smoothly one-hot encode them
train_targets = npz['targets'].astype(np.int)

#loading the validation data
npz = np.load('Employee_attrition_data_validation.npz')
validation_inputs, validation_targets = npz['inputs'].astype(np.float), npz['targets'].astype(np.int)

#loading the test data
npz = np.load('Employee_attrition_data_test.npz')
test_inputs, test_targets = npz['inputs'].astype(np.float), npz['targets'].astype(np.int)

### Building the model

In [3]:
# Setting the output size given a known 9 input_size
output_size = 2

# Using the same hidden layer size for both hidden layers
hidden_layer_size = 50
    
# defining how the model looks
model = tf.keras.Sequential([
    
    # 1st hidden layer
    tf.keras.layers.Dense(hidden_layer_size, activation='relu'), 
    
    # 2nd hidden layer
    tf.keras.layers.Dense(hidden_layer_size, activation='relu'), 

    # output layer
    tf.keras.layers.Dense(output_size, activation='softmax') 
])


#Choosing the optimizer and the loss function

model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

#Training our built model

# setting the batch size
batch_size = 100

# setting a maximum number of training epochs
max_epochs = 100

# setting an early stopping mechanism to slightly tolerate the random validation loss increases by patience=2
early_stopping = tf.keras.callbacks.EarlyStopping(patience=2)

# fitting the model
model.fit(train_inputs, # train inputs
          train_targets, # train targets
          batch_size=batch_size, # batch size
          epochs=max_epochs, # epochs used if early stopping delays
          
          # callbacks called by a task when a checking for val_loss increases is completed
          callbacks=[early_stopping], # early stopping
          
          validation_data=(validation_inputs, validation_targets), # validation data
          
          verbose = 2 #ensures that our training generates enough information
          )  

Epoch 1/100
58/58 - 2s - loss: 0.5095 - accuracy: 0.7956 - val_loss: 0.3383 - val_accuracy: 0.8866
Epoch 2/100
58/58 - 0s - loss: 0.2940 - accuracy: 0.8920 - val_loss: 0.2374 - val_accuracy: 0.9188
Epoch 3/100
58/58 - 0s - loss: 0.2323 - accuracy: 0.9146 - val_loss: 0.2107 - val_accuracy: 0.9286
Epoch 4/100
58/58 - 0s - loss: 0.2144 - accuracy: 0.9214 - val_loss: 0.2007 - val_accuracy: 0.9356
Epoch 5/100
58/58 - 0s - loss: 0.2058 - accuracy: 0.9240 - val_loss: 0.1919 - val_accuracy: 0.9356
Epoch 6/100
58/58 - 0s - loss: 0.1973 - accuracy: 0.9274 - val_loss: 0.1902 - val_accuracy: 0.9342
Epoch 7/100
58/58 - 0s - loss: 0.1923 - accuracy: 0.9310 - val_loss: 0.1859 - val_accuracy: 0.9342
Epoch 8/100
58/58 - 0s - loss: 0.1874 - accuracy: 0.9337 - val_loss: 0.1832 - val_accuracy: 0.9370
Epoch 9/100
58/58 - 0s - loss: 0.1837 - accuracy: 0.9344 - val_loss: 0.1825 - val_accuracy: 0.9384
Epoch 10/100
58/58 - 0s - loss: 0.1801 - accuracy: 0.9356 - val_loss: 0.1768 - val_accuracy: 0.9384
Epoch 11/

### Testing the model

In [4]:
#variables to contain the two outputs from the evaluate function
#the default loss and whatever was specified in the 'metrics' argument when fitting the model
test_loss, test_accuracy = model.evaluate(test_inputs, test_targets)

23/23 [==============================] - 0s 1ms/step - loss: 0.1369 - accuracy: 0.9371


In [5]:
# Printing results
print('\nTest loss: {0:.2f}. Test accuracy: {1:.2f}%'.format(test_loss, test_accuracy*100.))


Test loss: 0.14. Test accuracy: 93.71%


***The 93.71% accuracy is good!!*** compared to val_accuracy of 94.54%  from the trained model

#### Observing predicted leaving employees from our test inputs

In [6]:
#getting the probability for an employee to leave
# In our case, we want ot know which COLUMN has the higher argument (probability), therefore we set axis=1 (for columns)
# The output would be the column ID with the highest probabilityfor each observation (row)
np.argmax(model.predict(test_inputs),axis=1)

array([0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0,
       0, 1, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0,
       1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1,
       0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1,
       0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0,
       0, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 1,
       1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 1,

The **1's indicate where the model has predicted a leaving employee** from our test inputs

In [7]:
#saving the model using the built-in TensorFlow method and .h5 extension optimal for large numerical objects
model.save('Employee_Attrition_model.h5') 